### PR TITLE
feat(python,rust): build_info() provides detailed information how polars was built

### LINKED ISSUE #5422: Provide detailed build information in the installed polars
### Problem description

When troubleshooting polars deployments it is quit useful to be able to access detailed informaiton about how this instance of `polars` was build. What features were enabled, what versions of dependencies used, what optimization debug flags, exact git commit hash, rustc version, etc.

A possible implementation calls for a function `polars.build_info()` that returns such information as a tree (a dict).

### PR DESCRIPTION
**Pull request by slonik-az**
(resolves #5422)

[new functionality is behind `build_info` feature gate]

`polars.build_info()` returns dictionary with hierarchically organized information about how this polars was built -- features enabled, dependencies and their versions, compilation details, optimization/debug flags, rustc version, git commit information, and more.

This information is useful when troubleshooting `polars` deployments.

### CODE REVIEW

File: py-polars/polars/show_versions.py [hunk 1]
```diff
@@ -69,3 +70,36 @@ def _get_dep_version(dep_name: str) -> str:
...
+        If compiled with "build_info" feature gate return the full build info,
+        otherwise only version is included.
+        """
+        return info
+
+    return build_info
+
+
+build_info = _closure_()
+del _closure_
```
Comment (zundertj) [inline py-polars/polars/show_versions.py:105]: Why would the method need to be deleted?

File: py-polars/tests/unit/test_show_versions.py [hunk 2]
```diff
@@ -8,3 +8,12 @@ def test_show_versions(capsys) -> None:  # type: ignore[no-untyped-def]
     assert "Python" in out
     assert "Polars" in out
     assert "Optional dependencies" in out
+
+
+def test_build_info() -> None:
+    build_info = pl.build_info()
+    features = build_info.get("features", {})
+    if features:
```
Comment (zundertj) [inline py-polars/tests/unit/test_show_versions.py:16]: It would be nice if we can test both branches here somehow.

File: py-polars/polars/show_versions.py [hunk 3]
```diff
@@ -69,3 +70,36 @@ def _get_dep_version(dep_name: str) -> str:
...
+
+        If compiled with "build_info" feature gate return the full build info,
+        otherwise only version is included.
+        """
+        return info
+
+    return build_info
+
+
+build_info = _closure_()
```
Comment (zundertj) [inline py-polars/polars/show_versions.py:104]: Is there a specific reason to run this on `import polars`? Why not leave as a method?

File: py-polars/polars/show_versions.py [hunk 4]
```diff
@@ -69,3 +70,36 @@ def _get_dep_version(dep_name: str) -> str:
 
     # all our dependencies (as of 2022-08-11) implement __version__
     return module_version
+
+
+def _closure_() -> _typing_.Callable[[], _typing_.Any]:
```
Comment (zundertj) [inline py-polars/polars/show_versions.py:75]: The method name could be more descriptive. I propose to call this `build_info`, and import this method directly in `__init__.py`

File: py-polars/polars/show_versions.py [hunk 5]
```diff
@@ -3,6 +3,7 @@
 import importlib
 import platform
 import sys
+import typing as _typing_
```
Comment (zundertj) [inline py-polars/polars/show_versions.py:6]: Why not do `from typing import Callable, Any`? The underscores are not needed either?

**Review by zundertj — CHANGES_REQUESTED** [review]
Left a couple of code-specific questions and suggestions (Python only).

On a high level, the most important question: why not integrate this information in `pl.show_versions`? 

And could you add an example output here (and in the docstring), so we know what we are talking about?

File: py-polars/polars/show_versions.py (hunk 1 above)
Comment (slonik-az) [reply py-polars/polars/show_versions.py:105]: `_closure_` should be called only once. After it is deleted it cannot be called at all. Otherwise a user can call it again as 
```python
import polars.show_versions as foo
foo._closure_()
```
Python unfortunately does not have proper private visibility. Closures is the only way to emulate privaye.

File: py-polars/polars/show_versions.py (hunk 3 above)
Comment (slonik-az) [reply py-polars/polars/show_versions.py:104]: It is a function at the same level as `show_versions` and `version`. If changing to method, of what class?

File: py-polars/polars/show_versions.py (hunk 4 above)
Comment (slonik-az) [reply py-polars/polars/show_versions.py:75]: The function is called `build_info`. Closure `_closure_` is here to create a private scope and initiate private variables that `build_info()` returns. Closure executed once and then deleted. This is standard way to overcome python's limitation of not having proper private or static variables.

File: py-polars/tests/unit/test_show_versions.py (hunk 2 above)
Comment (slonik-az) [reply py-polars/tests/unit/test_show_versions.py:16]: I think the code already tests both branches:

1. When `build_info` feature gate in enabled `features` field is populated and tested for presence of `BUILD_INFO` feature
2. Otherwise without `build_info` feature the only element present in dictionary is `version` and its presence is tested in the `else` part of the code (line 17, not shown in the snippet).

File: py-polars/polars/show_versions.py (hunk 5 above)
Comment (slonik-az) [reply py-polars/polars/show_versions.py:6]: Good call. I will change it and import only specific elements of `typing`.

**Comment by slonik-az** [comment]
> On a high level, the most important question: why not integrate this information in `pl.show_versions`?

`pl.show_versions` is already quite heavy and does two different things at once. (1) it prints some information to the stdout and (2) it populates and returns a dictionary with info on other python modules. Making it triple-duty loaded with rust build-info will make output (or/and return dictionary) huge and hard to parse. 

Also, `pl.build_info()` is aimed at use in logging and deployment troubleshooting. The fact that `pl.show_versions()` is indiscriminately prints to stdout (without possibility to redirect output) makes it ill-suited for this use case. `pl.build_info()` does one thing only (return dictionary with rust build information) leaving it up to user how and what to process.

To avoid confusion with `show_versions()`,  I can move `build_info()` into its own file `build_info.py`. The same for the tests.

**Comment by slonik-az** [comment]
> And could you add an example output here (and in the docstring), so we know what we are talking about?

Good call. I will show structure of the output in the docstring. The returned dictionary is large (it lists all 200+ polars rust dependencies and their versions). For brevity, I will include partial output.

**Comment by slonik-az** [comment]
Refactored python code per code review. Many thanks to @zundertj for insightful comments/suggestions. Also rebased the PR onto the latest master.

File: py-polars/polars/show_versions.py (hunk 1 above)
Comment (slonik-az) [reply py-polars/polars/show_versions.py]: I decided to refactor the code. `_closure_` is gone altogether and code is simplified.

`def build_info():` moved from `polars/show_versions.py` into `polars/build_info.py` to avoid any confusion with `show_versions()`.

File: py-polars/tests/unit/test_show_versions.py (hunk 2 above)
Comment (slonik-az) [reply py-polars/tests/unit/test_show_versions.py]: Moved testing code into `test_build_info.py` file.

File: py-polars/polars/show_versions.py (hunk 4 above)
Comment (slonik-az) [reply py-polars/polars/show_versions.py]: `_closure_()` is gone now. See the updated PR.

**Comment by zundertj** [comment]
@slonik-az : one thing I should have mentioned: could you also add an entry to this function in the Python API reference? I know this one is merged, but in a new PR.

**Comment by ritchie46** [comment]
Thanks @slonik-az for the PR and thank you @zundertj  for the review.

**Comment by slonik-az** [comment]
> could you also add an entry to this function in the Python API reference? I know this one is merged, but in a new PR.

Will do.

**Comment by slonik-az** [comment]
> could you also add an entry to this function in the Python API reference? I know this one is merged, but in a new PR.

Done in PR #5442
